### PR TITLE
fix(hooks): cap concurrent augment subprocesses (#1486)

### LINKED ISSUE #1486: GitNexus Hook Spawns Excessive Processes (~180 instances) Causing Severe System Load
### Area

gitnexus (CLI / core / indexing / MCP server)

### Summary

The GitNexus hook (`gitnexus-hook.cjs`) is being launched multiple times simultaneously, once approximately every day, causing significant system resource exhaustion. 

## Current Behavior

The process tree shows a cascade of nested shell executions launching the same GitNexus hook:

```
/bin/bash -lc node "gitnexus-hook.cjs"
  └─ /bin/sh -c /bin/bash -lc 'node "gitnexus-hook.cjs"'
      └─ /bin/bash -lc node "gitnexus-hook.cjs"
          └─ (repeats)
```

**Observed metrics:**
- **Process count:** 180+ instances of `gitnexus-hook.cjs` related processes
- **System load average:** 109.55, 57.41, 23.83 (extremely high)
- **Frequency:** Approximately once per day
- **Uptime impact:** System has been running for 7 days, 4:20

## Expected Behavior

Only one instance of the GitNexus hook should be running at any given time. The hook should execute and terminate cleanly without spawning nested processes.

## Reproduction Steps

1. Run Claude or any process that triggers the GitNexus hook
2. Monitor process tree: `ps axfu | grep nexus`
3. Observe accumulation of processes over time
4. Issue occurs approximately once daily

## Environment

- **User:** vlad
- **Hook path:** `/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs`
- **System:** Ubuntu (GNOME session)
- **Process pattern:** Recursive `/bin/sh` → `/bin/bash` → node execution

## Root Cause Analysis (Suspected)

The hook appears to be recursively calling itself, likely due to:
1. The hook triggering on events that cause it to re-trigger itself
2. Missing process locking mechanism (pidfile/semaphore)
3. Potential infinite loop condition triggered once daily

## Suggested Fix

1. **Implement process locking:** Use pidfile or flock to prevent multiple instances
2. **Add recursion detection:** Prevent the hook from triggering on its own operations
3. **Fix the root trigger:** Identify what causes the daily cascade and prevent the recursive spawning pattern
4. **Add timeout/circuit breaker:** Kill processes that exceed expected runtime

## Immediate Workaround

```bash
# Kill all runaway processes
pkill -f gitnexus-hook.cjs

# Temporary disable the hook
chmod -x /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
```

## Additional Context

The nested execution pattern shows `sh -c` spawning `bash -lc` which spawns another `sh -c`, creating a deep process chain. This suggests the hook may be:
- Executing a command that itself invokes the hook
- Using incorrect quoting/escaping in shell commands
- Missing proper process termination handling

### Context

_No response_

### Expected behavior

run with no issues

### Actual behavior

x

### Steps to reproduce

xx

### Environment

_No response_

### Logs / screenshots

```
vlad     1637484  0.0  0.0   2808  1844 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1637486  1.7  0.0 1271388 46496 ?       Sl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1640329 23.2  0.1 1596748 85876 ?       Rl   21:34   0:00  |   |           |       \_ npm exec gitnexus augment -- import.*arima
vlad     1639908  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1639911  0.8  0.0  12612  6588 ?        R    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1639934  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1639935  1.7  0.0 1137908 31660 ?       Rl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1639943  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1639944  3.2  0.0 1336912 46540 ?       Sl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1646322  0.0  0.0 1403944 48824 ?       Rl   21:34   0:00  |   |           |       \_ node /home/vlad/.nvm/versions/node/v24.11.1/bin/npx -y gitnexus augment -- arima
vlad     1639974  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1639981  3.3  0.0 1336924 46492 ?       Sl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1646060 85.7  0.1 1404004 49744 ?       Rl   21:34   0:00  |   |           |       \_ node /home/vlad/.nvm/versions/node/v24.11.1/bin/npx -y gitnexus augment -- forecast
vlad     1640034  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640040  3.3  0.0 1337168 46548 ?       Sl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1640108  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640110  3.1  0.0 1336884 45788 ?       Rl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1640144  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640151  0.6  0.0  12612  6588 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1640204  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640213  1.6  0.0 1137592 31852 ?       Rl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1640256  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640263  0.6  0.0  12612  6528 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1640305  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640310  1.9  0.0 1137848 32928 ?       Rl   21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1640356  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640365  1.0  0.0  12612  6624 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1640404  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640411  1.3  0.0 107012 17404 ?        R    21:34   0:00  |   |           |   \_ node /home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs
vlad     1640517  0.0  0.0   2808  1848 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640521  0.6  0.0  12612  6580 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1640678  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640692  0.7  0.0  12612  6624 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1640883  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1640898  0.7  0.0  12612  6616 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1641157  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1641160  1.1  0.0  12612  6600 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1641325  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1641345  1.2  0.0  12612  6608 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1641566  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1641576  1.2  0.0  12612  6612 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1641823  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1641833  0.8  0.0  12612  6572 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642010  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642033  0.9  0.0  12612  6584 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642105  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642120  0.4  0.0  12612  6584 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642283  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642321  1.4  0.0  12612  6628 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642389  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642399  1.0  0.0  12612  6616 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642484  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642502  1.0  0.0  12612  6592 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642690  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642695  1.0  0.0  12612  6628 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1642936  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1642941  1.1  0.0  12612  6580 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1643195  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1643206  1.7  0.0  12612  6628 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1643380  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1643386  1.2  0.0  12612  6604 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1643647  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1643652  2.1  0.0  12612  6576 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1643838  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1643840  2.3  0.0  12612  6604 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1643931  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1643939  1.6  0.0  12612  6560 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644095  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1644100  0.9  0.0  12612  6508 ?        R    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644265  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1644277  1.9  0.0  12612  6568 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644937  0.0  0.0  12612  4816 ?        S    21:34   0:00  |   |           |       \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644953  0.0  0.0  12612  4496 ?        S    21:34   0:00  |   |           |           \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644954  0.0  0.0  12744  5080 ?        S    21:34   0:00  |   |           |               \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644596  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1644613  2.3  0.0  12612  6552 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645092  0.0  0.0  12612  4804 ?        S    21:34   0:00  |   |           |       \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645115  0.0  0.0  12612  4612 ?        S    21:34   0:00  |   |           |           \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645118  0.0  0.0  12744  5072 ?        S    21:34   0:00  |   |           |               \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1646222  0.0  0.0  12744  4688 ?        S    21:34   0:00  |   |           |               |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1646223  0.0  0.0  12744  5332 ?        S    21:34   0:00  |   |           |               |       \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1644911  0.0  0.0   2808  1852 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1644916  2.7  0.0  12612  6496 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645383  0.0  0.0  12612  4808 ?        S    21:34   0:00  |   |           |       \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645402  0.0  0.0  12612  4616 ?        S    21:34   0:00  |   |           |           \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645403  0.0  0.0  12744  5076 ?        S    21:34   0:00  |   |           |               \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645038  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1645040  2.9  0.0  12612  6544 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645555  0.0  0.0  12612  4856 ?        S    21:34   0:00  |   |           |       \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645600  0.0  0.0  12612  4600 ?        S    21:34   0:00  |   |           |           \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645603  0.0  0.0  12744  5128 ?        S    21:34   0:00  |   |           |               \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645179  0.0  0.0   2808  1856 ?        Ss   21:34   0:00  |   |           \_ /bin/sh -c /bin/bash -lc 'node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"'
vlad     1645187  3.2  0.0  12612  6556 ?        S    21:34   0:00  |   |           |   \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645733  0.0  0.0  12612  4812 ?        S    21:34   0:00  |   |           |       \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645748  0.0  0.0  12612  4492 ?        S    21:34   0:00  |   |           |           \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1645749  0.0  0.0  12744  5080 ?        S    21:34   0:00  |   |           |               \_ /bin/bash -lc node "/home/vlad/.claude/hooks/gitnexus/gitnexus-hook.cjs"
vlad     1646268  0.0  0.0   9152  2328 pts/12   S+   21:34   0:00  |       \_ grep --color=auto nexus
⋊> ~/w/l/fk-bigquery-queries on feature/fk-import-service ⨯ ps axfu | grep nexus | wc -l                                                                                        21:34:42
180
⋊> ~/w/l/fk-bigquery-queries on feature/fk-import-service ⨯ w                                                                                                                   21:34:48
 21:34:49 up 7 days,  4:20,  1 user,  load average: 109.55, 57.41, 23.83
USER     TTY      FROM             LOGIN@   IDLE   JCPU   PCPU  WHAT
vlad     tty2     -                03May26  7days  0.01s   ?    /usr/libexec/gnome-session-binary --session=ubuntu

```

### PR DESCRIPTION
**Pull request by abhigyanpatwari**
## Summary

Fixes #1486. The `gitnexus-hook.cjs` PreToolUse hook had no cap on concurrent `gitnexus augment` subprocesses. When Claude Code fired parallel Grep/Glob/Bash tool calls, each spawned its own Node + LadybugDB cold start, holding resources for several seconds. Under heavy parallel-search bursts these accumulated faster than they finished — the reporter saw 180+ piled-up processes and a load average over 100.

### Diagnosis (not what the report claims)

The report blames "recursive self-invocation" in the hook. **The hook does not invoke itself.** Every `spawnSync` call in `gitnexus-hook.cjs` is shell-free and direct — `git`, `node <cliPath>`, `npx` — none invoke `bash`. The nested `bash -lc → bash -lc → bash -lc` chain in the reporter's process tree is produced by Claude Code's `sh -c '<command>'` wrapper combined with the user's login-shell init scripts (likely nvm / shell-integration) re-exec'ing. That isn't something the hook can or should fix.

The real bug — and what causes the same observable symptoms — is **the absence of a concurrency guard**. N parallel tool calls = N concurrent `gitnexus augment` cold-starts, with nothing capping the fan-out.

### Fix

Add a lockfile-based concurrency guard under `<.gitnexus>/.hook-locks/`:

- Each running hook writes `<pid>.lock` before spawning augment.
- Before claiming a slot, the guard counts existing lockfiles, pruning entries whose PID is no longer alive (`process.kill(pid, 0)`) or whose mtime is older than 30s (safety net against PID reuse / hung hooks).
- If `MAX_INFLIGHT` (3) live slots are already in use, the hook bails silently. Augment is best-effort enrichment — missing a few fires during a burst is far better than melting the box.
- Cleanup runs in `finally {}` and via `process.on('exit')` for hard kills.

Applied to all three hook variants that spawn augment:
- `gitnexus/hooks/claude/gitnexus-hook.cjs` (npm-installed Claude hook copied by `gitnexus setup`)
- `gitnexus-claude-plugin/hooks/gitnexus-hook.js` (Claude plugin hook)
- `gitnexus-cursor-integration/hooks/gitnexus-hook.cjs` (Cursor hook — same shape, same exposure)

### Out of scope

- The 7s `spawnSync` timeout is unchanged. Tightening it without telemetry on real cold-start tails risks breaking legitimate slow-but-completing augments.
- The `npx -y gitnexus` fallback cold start remains heavy. A longer-term fix would route augment through the already-running MCP server (overlaps with #1492's lock-contention angle).
- The user's nested `bash -lc` shell wrapping is environment-specific and upstream of GitNexus.

## Test plan

- [x] `npx vitest run test/unit/hooks.test.ts` — all 95 tests pass (4 new)
- [x] `npx vitest run test/unit/cursor-hook.test.ts` — 41 tests pass
- [x] `npx tsc --noEmit` — clean
- [x] New regression tests cover:
  - source-level: `acquireHookSlot` is defined and called from `handlePreToolUse` in both CJS + Plugin variants
  - integration: hook bails silently with 3 sentinel lockfiles for genuinely live PIDs (spawned `node -e "setTimeout(...)"` children)
  - integration: hook prunes lockfiles for dead PIDs

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adds a new cross-process lockfile-based concurrency cap around `gitnexus augment`, which can change when/if augmentation runs under heavy parallel tool usage and may surface edge cases on unusual filesystems or permissions.
> 
> **Overview**
> PreToolUse/postToolUse hooks now wrap `gitnexus augment` with a **hard concurrency guard** that limits in-flight subprocesses to 3 via atomic fixed-name lock slots under `.gitnexus/.hook-locks/`, including stale lock takeover based on PID liveness and age.
> 
> When the lock directory can’t be created or all slots are occupied, the hook **skips augmentation silently** instead of spawning additional processes, and the new unit/integration tests assert the guard wiring, fail-closed behavior, slot reclamation, and the hard-cap under burst contention.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 482ce524581346d50da6504bb17c55be9506e9cc. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by abhigyanpatwari** [comment]
@claude review it

File: gitnexus/hooks/claude/gitnexus-hook.cjs [hunk 1]
```diff
@@ -160,6 +160,128 @@ function extractPattern(toolName, toolInput) {
...
+            isLive = true;
+          } else {
+            const owner = Number.parseInt(ownerStr, 10);
+            if (Number.isFinite(owner) && owner > 0) {
+              try {
+                process.kill(owner, 0);
+                isLive = true;
+              } catch {
+                /* ESRCH (or EPERM under cross-user) — treat as dead */
+              }
```
[the remark this answers is not in the document]
Comment (abhigyanpatwari) [reply gitnexus/hooks/claude/gitnexus-hook.cjs]: Already addressed — commit [5bacf5a4](https://github.com/abhigyanpatwari/GitNexus/commit/5bacf5a4) (now on PR HEAD) distinguishes `e.code === 'ESRCH'` (dead → reclaim) from any other error including `EPERM` (alive → keep). All three hook variants got the same fix.

```js
} catch (e) {
  if (e && e.code === 'ESRCH') isLive = false;
  else isLive = true; // EPERM or unknown → conservatively alive
}
```

**Comment by abhigyanpatwari** [comment]
Finalising — pushed commit [`482ce524`](https://github.com/abhigyanpatwari/GitNexus/commit/482ce524). Status of the open review threads:

| Finding | Source | Status |
| --- | --- | --- |
| EPERM vs ESRCH in PID liveness check | [Cursor Bugbot](https://github.com/abhigyanpatwari/GitNexus/pull/1510#discussion_r3227095897) | Already addressed in [`5bacf5a4`](https://github.com/abhigyanpatwari/GitNexus/commit/5bacf5a4) — `if (e.code === 'ESRCH') isLive=false; else isLive=true` covers EPERM and unknown errors. |
| Lock-dir creation failure proceeds unguarded | Production-readiness review (Finding 1, Medium) | Fixed in `482ce524` — `mkdirSync` catch now returns `null` so the caller's `if (!release) return;` skips augment cleanly. Fail-open would have reintroduced the #1486 fan-out on read-only / cross-user `.gitnexus/` setups. |
| "PID-liveness wins over age" comment is misleading | Production-readiness review (Finding 2, Low) | Fixed in `482ce524` — comment now states that liveness wins below `HOOK_LOCK_STALE_MS`, age is the final arbiter beyond 30s for PID-reuse defence. No code change; the existing branch is already correct. |
| Cursor variant missing 10-spawner burst test | Production-readiness review (Finding 3, Low) | Algorithm is byte-for-byte identical to the CJS hook (already covered by the burst test in `hooks.test.ts`). Added an inline comment to `cursor-hook.test.ts` documenting this so future maintainers don't read it as a gap. |
| CodeQL file-system race | github-advanced-security | Single-fd `openSync → fstatSync → readSync → closeSync` in [`e11077b3`](https://github.com/abhigyanpatwari/GitNexus/commit/e11077b3) — CodeQL JS/TS analysis now passes. |

Also added two source-level regression tests asserting `acquireHookSlot` returns `null` (not `() => {}`) on lock-dir failure, in both `hooks.test.ts` (CJS + Plugin) and `cursor-hook.test.ts`.

`npx vitest run test/unit/hooks.test.ts test/unit/cursor-hook.test.ts` → 148/148 pass locally. `npx tsc --noEmit` clean. Ready to merge once CI re-runs on `482ce524`.

File: gitnexus/hooks/claude/gitnexus-hook.cjs [hunk 2]
```diff
@@ -160,6 +160,135 @@ function extractPattern(toolName, toolInput) {
...
+        } catch {
+          /* another hook beat us to it — retry will hit EEXIST */
+        }
+        // Loop and retry this slot.
+      }
+    }
+  }
+
+  return null;
+}
```
[the remark this answers is not in the document]
Comment (magyargergo) [reply gitnexus/hooks/claude/gitnexus-hook.cjs]: @copilot address this finding and follow DoD.md

**Comment by magyargergo** [comment]
@copilot address the following findings and follow DoD.md file

~100-line locking function triplicated across three hook files

Low Severity

The acquireHookSlot function and its three constants (HOOK_LOCK_SUBDIR, HOOK_LOCK_MAX_INFLIGHT, HOOK_LOCK_STALE_MS) are copy-pasted identically across all three hook files — roughly 100 lines of non-trivial locking logic with PID-liveness checks, stale-takeover, and fd-based TOCTOU avoidance. A future bug fix or behavioral change to the concurrency guard needs to be synchronized across all three copies, with real risk of divergence.

Additional Locations (2)
gitnexus-claude-plugin/hooks/gitnexus-hook.js#L186-L290
gitnexus-cursor-integration/hooks/gitnexus-hook.cjs#L216-L320
